File: proteomics/analysis/deg_analysis/R/volcano_plot.py
import re
from pathlib import Path

from proteomics.analysis.deg_analysis.base_args import (
    RDegPlotArgs,
    RConfig,
    RunRDegAnalysis,
)


class VolcanoArgs(RDegPlotArgs):
    upregulated_color: str = "red"
    downregulated_color: str = "royalblue"
    not_sig_color: str = "black"


class VolcanoPlot(VolcanoArgs, RunRDegAnalysis[Path]):
    def get_r_script(self) -> Path:
        return Path(__file__).parent / "volcano-plot.R"

    def process_stdout(self, stdout: str) -> Path | None:
        try:
            cleaned_string = re.sub(r"^\[\d+]\s*", "", stdout[-1].strip())
            volcano_png = Path(cleaned_string.strip('"'))
            # make sure the file exists
            if not volcano_png.exists() or not volcano_png.suffix == ".png":
                raise FileNotFoundError(
                    f"File {volcano_png} is not a png file"
                )
            # make sure its a png file
        except FileNotFoundError as e:
            print(f"Error: {e}")
            volcano_png = None

        return volcano_png


def run_volcano_plot_r(
    *,
    r_config: RConfig,
    output_dir: Path,
    deg_results: Path,
    experiment: str,
    volcano_args: VolcanoArgs,
) -> Path | None:
    volcano_plot = VolcanoPlot(
        output_dir=output_dir,
        input_file=deg_results,
        rscript_bin=r_config.rscript_bin,
        experiment=experiment,
        **volcano_args.model_dump(),
    )

    return volcano_plot.run_analysis()


def main():
    r_config = RConfig(
        rscript_bin="/Users/ac4294/.pyenv/versions/miniforge3-24.11.3-0/envs/liver_prot_s2808d_2025/bin/Rscript"
    )
    output_dir = Path("/Users/ac4294/Downloads")
    deg_results = Path(
        "/Users/ac4294/dev/2025/yang_liver_proteomics_s2808d/data/example/_results-1742436624597126/S2808D_HFD_vs_WT_HFD/limma_outputs/S2808D_HFD_vs_WT_HFD_deg_limma.csv"
    )
    experiment = "S2808D_HFD_vs_WT_HFD"
    volcano_args = VolcanoArgs(
        gene_column="X",
        lfc_column="logFC",
        pval_column="adj.P.Val",
        fc_threshold=1.5,
        pval_threshold=0.05,
        width=6,
        height=8,
    )

    enrichment_results = run_volcano_plot_r(
        r_config=r_config,
        output_dir=output_dir,
        deg_results=deg_results,
        experiment=experiment,
        volcano_args=volcano_args,
    )

    print("The volcano plot results")
    print(enrichment_results)


if __name__ == "__main__":
    main()

File: proteomics/analysis/deg_analysis/R/volcano-plot.R
library(ggplot2)
library(EnhancedVolcano)
# Load the necessary library
library(gridExtra)
library(optparse)
source("base_args.R")

get_args <- function() {
  option_list <- list(
    make_option(
      c("--upregulated_color"),
      type = "character",
      default = "red",
      help = "The color for upregulated genes. Default is red"
    ),
    make_option(
      c("--downregulated_color"),
      type = "character",
      default = "royalblue",
      help = "The color for downregulated genes. Default is royalblue"
    ),
    make_option(
      c("--not_sig_color"),
      type = "character",
      default = "black",
      help = "The color for non-significant genes. Default is black"
    )
  )

  base_args_list <- get_base_args()
  # merge the two lists
  option_list <- c(base_args_list, option_list)

  # Create OptionParser object
  opt <- parse_args(OptionParser(option_list = option_list))

  return(opt)
}


run_volcano <- function(
  input_file,
  output_dir,
  genes_column = "X",
  # the column name for the gene names from Limma. By default its blank (X)
  log2_fc_col = "logFC",
  p_val_col = "adj.P.Val",
  p_cutoff = 0.05,
  fc_cutoff = 1.5,
  upregulated_color = "red",
  downregulated_color = "royalblue",
  not_sig_color = "black",
  plot_title = "Volcano Plot"
) {

  lfc_cutoff <- log2(fc_cutoff)

  # Load in the data (limma results)
  lfc_data <- read.csv(input_file)

  # Create a dataframe from the data and set the row names to the Gene column
  data_df <- as.data.frame(lfc_data, row.names = lfc_data[[genes_column]])

  num_de <- nrow(
    # must be both greater than the lfc_cutoff and less than the p_cutoff
    data_df[which(abs(data_df[[log2_fc_col]]) > lfc_cutoff & data_df[[p_val_col]] < p_cutoff),]
  )

  # create a colormap
  keyvals <- ifelse(
    data_df[[log2_fc_col]] < -lfc_cutoff & data_df[[p_val_col]] < p_cutoff, downregulated_color,
    ifelse(data_df[[log2_fc_col]] > lfc_cutoff & data_df[[p_val_col]] < p_cutoff, upregulated_color,
           not_sig_color
    )
  )

  caption <- paste0("Total: ", nrow(data_df), ". Significant: ", num_de)

  print(caption)

  keyvals[is.na(keyvals)] <- not_sig_color
  names(keyvals)[keyvals == upregulated_color] <- "Up"
  names(keyvals)[keyvals == not_sig_color] <- "Not significant"
  names(keyvals)[keyvals == downregulated_color] <- "Down"

  print("Creating plot")

  plot <- EnhancedVolcano(
    data_df,
    lab = rownames(data_df),
    selectLab = c(""),
    drawConnectors = TRUE,
    widthConnectors = 0.5,
    x = log2_fc_col,
    y = p_val_col,
    title = plot_title,
    FCcutoff = lfc_cutoff,
    pCutoff = p_cutoff,
    caption = caption,
    subtitle = "",
    borderWidth = 1.0,
    borderColour = "black",
    border = "full",
    colCustom = keyvals,
    subtitleLabSize = -1,
    legendPosition = "right",
  )

  # Set the limits of the plot
  maxy <- ceiling(-log10(min(data_df[[p_val_col]], na.rm = TRUE)))
  minx <- floor(min(data_df[[log2_fc_col]], na.rm = TRUE))
  maxx <- ceiling(max(data_df[[log2_fc_col]], na.rm = TRUE))
  plot <- plot + coord_cartesian(ylim = c(NA, maxy), xlim = c(minx, maxx))

  print("Saving plot")
  for (format in c("svg", "pdf", "png")) {
    plot_name <- file.path(output_dir, paste0("volcano_plot.", format))
    ggsave(plot_name, plot, dpi = 400, width = 9, height = 8.5)
  }

  print(
    file.path(output_dir, paste0("volcano_plot.", format))
  )
}

main <- function() {
  args <- get_args()

  if (!dir.exists(args$output_dir)) {
    stop("Output directory does not exist")
  }

  run_volcano(
    input_file = args$input_file,
    output_dir = args$output_dir,
    genes_column = args$gene_column,
    log2_fc_col = args$lfc_column,
    p_val_col = args$pval_column,
    fc_cutoff = args$fc_threshold,
    p_cutoff = args$pval_threshold,
    upregulated_color = args$upregulated_color,
    downregulated_color = args$downregulated_color,
    not_sig_color = args$not_sig_color,
    plot_title = args$experiment
  )
}

main()
